snowfall/training/mmi_graph.py

```diff
@@ -110,49 +109,194 @@ def compile(self,
+                # k2.invert() is not an identity map
+                HPL_inv, hpl_inv_to_hpl = k2.invert(HPL, ret_arc_map=True)
+
+                HPL_inv_sorted, hpl_inv_sorted_to_hpl_inv = k2.arc_sort(HPL_inv, ret_arc_map=True)
+
+                hpl_inv_sorted_to_hpl = k2.compose_arc_maps(hpl_inv_to_hpl, hpl_inv_sorted_to_hpl_inv)
+
+                hpl_inv_sorted_to_hp_inv_sorted = k2.compose_arc_maps(hpl_to_hp_inv_sorted, hpl_inv_sorted_to_hpl)
+
+                hpl_inv_sorted_to_hp = k2.compose_arc_maps(hp_inv_sorted_to_hp, hpl_inv_sorted_to_hp_inv_sorted)
+
+                hpl_inv_sorted_to_p_self_loops = k2.compose_arc_maps(hp_to_p_self_loops, hpl_inv_sorted_to_hp)
+
+                hpl_inv_sorted_to_p = k2.compose_arc_maps(p_self_loops_to_p, hpl_inv_sorted_to_p_self_loops)
+
+                # for the numerator graph
+                self.HPL_inv_sorted = HPL_inv_sorted
+                self.hpl_inv_sorted_to_p = hpl_inv_sorted_to_p
+                self.lm_scores = self.HPL_inv_sorted.scores.clone()
 
-        ctc_topo_P = k2.arc_sort(ctc_topo_P)
+                # for the denominator graph
+                hp_to_p = k2.compose_arc_maps(p_self_loops_to_p, hp_to_p_self_loops)
 
-        num_graphs = self.build_num_graphs(texts)
-        num_graphs_with_self_loops = k2.remove_epsilon_and_add_self_loops(
-            num_graphs)
+                self.HP = HP
+                assert self.HP.scores.abs().sum().item() == 0
+                self.hp_to_p = hp_to_p
 
-        num_graphs_with_self_loops = k2.arc_sort(num_graphs_with_self_loops)
+                assert self.HPL_inv_sorted.requires_grad is False
+                assert self.HP.requires_grad is False
+                assert self.lm_scores.requires_grad is False
 
-        num = k2.compose(ctc_topo_P,
-                         num_graphs_with_self_loops,
-                         treat_epsilons_specially=False)
-        num = k2.arc_sort(num)
+        self.HPL_inv_sorted.scores = self.lm_scores + k2.index(P.scores, self.hpl_inv_sorted_to_p)
+
+        # When this function is executed inside `with torch.no_grad()` context,
+        # The following assert will fail.
+        #  assert self.HPL_inv_sorted.requires_grad == P.requires_grad
+
+        linear_fsas = self.build_linear_fsas(texts)
+        linear_fsas_with_self_loops = k2.add_epsilon_self_loops(linear_fsas)
+
+        b_to_a_map = torch.zeros(len(texts),
+                                 dtype=torch.int32,
+                                 device=self.device)
+
+        num_graphs = k2.intersect_device(self.HPL_inv_sorted,
+                                         linear_fsas_with_self_loops,
+                                         b_to_a_map,
+                                         sorted_match_a=True)
+        num_graphs = k2.invert(num_graphs)
+
+        # TODO(fangjun): k2.connect supports only CPU.
+        # Add CUDA support.
+        num_graphs = k2.connect(num_graphs.to('cpu')).to(P.device)
```

Comment:
@danpovey 

I think we probably need a CUDA version of k2.connect(),
though I have not profiled this pull-request. It is currently
slower than before. Not sure if it is the problem of TaskRedirect or is caused by this statement.

Reply:
is .connect() really necessary?

Reply:
> is .connect() really necessary?

If I don't invoke `k2.connect`, then the resulting `get_tot_scores` for the num_lats returns all `-inf`.
If `k2.connect` is used, then `get_tot_scores` returns no `-inf`s.

Reply:
Something is not right here.
I think it may be a mistake to compose ctc_topo unless it's right at the end.  I believe ctc_topo expects to be composed with something that was epsilon-free and which then had epsilon self-loops added.  Because we are interpreting the epsilons on one side as "blank", which is in a sense a real symbol, things are a little subtle there.

Reply:
... so I think it may be OK to compose L and P, and to compose that with the transcripts, but I'd leave ctc_topo until the end.

Reply:
> I think it may be a mistake to compose ctc_topo unless it's right at the end. I believe ctc_topo expects to be composed with something that was epsilon-free and which then had epsilon self-loops added.

I am using `intersect_device(ctc_topo_inv, P_with_self_loops).invert()` (equivalent to `compose(ctc_topo, P, treat_epslion_speciall=True)`.

There is no `0` (neither blank nor epsilons) in P, so I think it is correct.